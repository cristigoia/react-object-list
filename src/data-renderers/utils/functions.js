/**
 * Gets the value of the cell from the row and keys provided
 * @param  {Object} row  Row supplied
 * @param  {Array|string} keys List of keys used to access the value
 * @return {(Object|string|number)}      Value from row, or row itself if no keys
 */
const getValueFromAccessor = (row, keys) => {
  if (!Array.isArray(keys)) keys = keys.split('.')
  let value = row
  const cloneKeys = [...keys]
  while (value && cloneKeys.length) {
    const thisKey = cloneKeys.shift()
    if (thisKey in value) {
      value = value[thisKey]
    }
  }
  return value
}

export {
  getValueFromAccessor,
}
