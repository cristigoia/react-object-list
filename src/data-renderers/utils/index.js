export {
  getValueFromAccessor,
  handleRowClick,
} from './functions'
