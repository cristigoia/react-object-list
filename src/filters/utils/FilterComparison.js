import React from 'react'
import PropTypes from 'prop-types'
import Select from 'react-select'

class FilterComparison extends React.Component {
  static propTypes = {
    options: PropTypes.arrayOf(PropTypes.shape({
      label: PropTypes.string,
      value: Select.propTypes.value,
    })),
    value: Select.propTypes.value,
    onChange: PropTypes.func.isRequired,
  }

  render() {
    const {options, value} = this.props
    return (
      <div className="objectlist-current-filter__filter-comparison">
        <Select
          options={options}
          value={value}
          onChange={newValue => this.props.onChange(newValue.value)}
          clearable={false}
        />
      </div>
    )
  }
}

export default FilterComparison
