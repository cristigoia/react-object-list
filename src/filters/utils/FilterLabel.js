import React from 'react'
import PropTypes from 'prop-types'

/**
 * Component to render label for filters
 */
class FilterLabel extends React.Component {
  static propTypes = {
    /** Text to render in the label */
    label: PropTypes.string.isRequired,
  }
  render() {
    return (
      <label className="objectlist-current-filter__filter-name">{this.props.label}:</label>
    )
  }
}

export default FilterLabel
