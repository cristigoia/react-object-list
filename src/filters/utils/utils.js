const possibleQueryStringLookupExpressions = ['exact', 'lt', 'gt', 'gte', 'lte']

export {
  possibleQueryStringLookupExpressions,
}
