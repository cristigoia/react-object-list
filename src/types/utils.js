function floatPrecision(floatValue, precision) {
  floatValue = parseFloat(floatValue)
  if (isNaN(floatValue)) { return parseFloat('0').toFixed(precision) } else {
    const power = Math.pow(10, precision)
    floatValue = (Math.round(floatValue * power) / power).toFixed(precision)
    return floatValue.toString()
  }
}

function commaNumber(numberString) {
  const parts = numberString.toString().split('.')
  parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ',')
  return parts.join('.')
}
/**
 * Converts a float to stringed currency representation
 * @param  {number} floatNumber Amount to convert
 * @return {string}             Dollar string
 */
export function currency(floatNumber) {
  return `${floatNumber < 0 ? '-' : ''}$${commaNumber(floatPrecision(Math.abs(floatNumber), 2))}`
}
