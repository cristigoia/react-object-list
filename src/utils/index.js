export * from './constants'
export * from './proptypes'
export * from './functions'
export loadingSpinner from './loading'
export ErrorMessage from './ErrorMessage'
