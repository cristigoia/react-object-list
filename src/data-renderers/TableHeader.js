import React from 'react'
import PropTypes from 'prop-types'
import HeaderField from './HeaderField'
import WidthHandle from './WidthHandle'
import {COLUMN_TYPE} from '../utils/proptypes'
import {valueEqual} from '../utils/functions'

/**
 * An individual header for a column containing a HeaderField for display
 * and an optional WidthHandle to resize the column width
 */
export default class TableHeader extends React.Component {
  static propTypes = {
    /**
     * the HeaderFields that is to be displayed within this TableHeader
     * array of objects eg. [{name: 'Property', sortKey: 'name', className: ''}]
     * when empty the other props will be used to generate a single HeaderField
     */
    headerItems: PropTypes.oneOfType([PropTypes.arrayOf(COLUMN_TYPE), COLUMN_TYPE]).isRequired,
    saveWidth: PropTypes.func,
    label: PropTypes.string,
    /** callback function for the objectlist to set the active sorting of the data */
    updateSorting: PropTypes.func,
    sortKeys: PropTypes.arrayOf(PropTypes.shape({
      sortKey: PropTypes.string,
      value: PropTypes.bool,
    })),
    /** class to apply to the th element of the header */
    className: PropTypes.string,
  }

  static defaultProps = {
    headerItems: [],
    className: '',
    sortKeys: [],
  }

  state = {
    headerItems: [],
  }

  static getDerivedStateFromProps = (props, state) => {
    const headerItems = Array.isArray(props.headerItems) ? props.headerItems : [props.headerItems]
    const label = props.label || headerItems[0].dataKey.split('.').pop()
    const width = (headerItems[0] || {}).width
    return ({headerItems, width, label})
  }

  shouldComponentUpdate(nextProps, nextState) {
    return (
      nextProps.label !== this.props.label ||
      nextProps.className !== this.props.className ||
      !valueEqual(this.props.sortKeys, nextProps.sortKeys) ||
      !valueEqual(this.state.headerItems, nextState.headerItems)
    )
  }

  /**
   * Callback function used to set the width for this column
   */
  setWidth = (width) => {
    this.setState(prevState => {
      const newWidth = prevState.width + parseInt(width)
      return {
        width: (newWidth >= 20) ? newWidth : prevState.width,
      }
    })
  }

  /**
   * Save the preferred column width to user preferences
   */
  saveWidth = () => {
    this.props.saveWidth(this.state.label, this.state.width)
  }

  /**
   *
   * @returns {bool} - true for ascending, false for descending, null for no sort
   */
  getSortDirection = key => {
    const direction = this.props.sortKeys.find(sortKey => sortKey.sortKey === key)
    return (direction === undefined) ? null : direction.value
  }

  render() {
    const headers = this.state.headerItems.map((header, i) => (
      <HeaderField
        key={`headerfield-${i}`}
        activeSort={this.getSortDirection(header.sortKey)}
        updateSorting={this.props.updateSorting}
        {...header}
      />
    ))
    let widthHandle
    if (this.state.width && this.props.saveWidth) {
      widthHandle = (
        <WidthHandle
          onChange={this.setWidth}
          onSave={this.saveWidth}
        />
      )
    }

    return (
      <th width={this.state.width} className={`objectlist-table__th ${this.props.className}`}>
        { headers }
        { widthHandle }
      </th>
    )
  }
}
