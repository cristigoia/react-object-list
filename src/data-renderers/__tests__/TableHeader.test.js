import React from 'react'
import { snapshotTest } from 'utils/tests'
import TableHeader from '../TableHeader'
jest.mock('../HeaderField', () => 'HeaderField')

describe('Table Header', () => {
  describe('Snapshots', () => {
    const headerItem = {
      dataKey: 'header_sort_key',
      header: 'Header Text',
      item: jest.fn(),
      sortable: true,
      optional: false,
    }

    it('renders correctly', () => {
      const props = {
        headerItems: {...headerItem},
        setSort: jest.fn(),
      }
      snapshotTest(<TableHeader {...props} />)
    })
    it('renders correctly with multiple header items', () => {
      const props = {
        headerItems: [{...headerItem}, {...headerItem, header: 'Header Two Text', key: 'header_two_key'}],
        setSort: jest.fn(),
      }
      snapshotTest(<TableHeader {...props} />)
    })
  })
})
