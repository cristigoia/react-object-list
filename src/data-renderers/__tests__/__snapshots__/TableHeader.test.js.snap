// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Header Snapshots renders correctly 1`] = `
<th
  className="objectlist-table__th "
  width={undefined}
>
  <HeaderField
    activeSort={null}
    dataKey="header_sort_key"
    header="Header Text"
    item={[MockFunction]}
    optional={false}
    sortKey="🍬"
    updateSorting={undefined}
  />
</th>
`;

exports[`Table Header Snapshots renders correctly with multiple header items 1`] = `
<th
  className="objectlist-table__th "
  width={undefined}
>
  <HeaderField
    activeSort={null}
    dataKey="header_sort_key"
    header="Header Text"
    item={[MockFunction]}
    optional={false}
    sortKey="🍬"
    updateSorting={undefined}
  />
</th>
`;
